### PR TITLE
Feature/migrate opensearch 1.2.0

### PR DESCRIPTION
**Pull request by RishiDiwanTT**
## Description
Only changes required were of _type removal due to single index - single type changes
https://www.elastic.co/guide/en/elasticsearch/reference/7.17/removal-of-types.html
Additionally the opensearch(1.2.0) was configured to remove the security plugin so that SSL and Basic Auth is not required
These changes create the need to reconfigure the search index as an Opensearch index

## Motivation and Context
As Elasticsearch 6.8 is EOL, the search index needed to be upgraded.
As AWS has stopped support for Elasticsearch, and will continue to support only Opensearch in its managed offering, we have chosen to continue down the path of opensearch

## How Has This Been Tested?

Unit tests have been updated and run.
Manually, the `search_index_clean` and `search_index_refresh` have been run
Manually, the `/search` API has been checked

## Checklist

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by jonathangreen — CHANGES_REQUESTED** [review]
I'm going to start with a high level review here, then spin up some docker containers to do some deeper testing, so there may be some more comments come up afterwords.

My main comment here, and the thing I'd like to see addressed before merging is to decouple the merging of this PR from the infrastructure work needed to support this change. That will allow us to keep running ES in old environments and allow us to phase in this change.

To that end here is the additional work I'd like to see on this PR: 
- Update this PR so the choice of running ES or OS is a configuration setting. You can either configure your site to work with Elasticsearch or Opensearch. 
  - This can be lightweight and klugey, the goal eventually will be to remove all the ES code, but for now I think its a better option to be able to run both. 
  - Make some notes with todos in the code about what can be removed when removing ES support.
- Add both ES and OS to tox (and github actions) so that our tests will run againt both
- Update our README and the docker/README file to reflect the change to OS and to note that ES is being deprecated and will be removed in a future version.

After that we can get this PR merged and then start slowly rolling it out in our test sites and then our client sites.

**Comment by jonathangreen** [comment]
A couple questions that come up as I'm going through this: 
- Do we need to update any libraries to be compatible with Opensearch? I see we are still using the elasticsearch libraries
- Why are we targeting Opensearch 1.2.0? It looks like the most recent Opensearch version is 1.3.5

**Comment by jonathangreen** [comment]
Just merged in main, so I can work with latest code while testing this branch

**Comment by jonathangreen** [comment]
So it looks like even though `main` merged in cleanly, some of the work we've done in:
https://github.com/ThePalaceProject/circulation/pull/453
https://github.com/ThePalaceProject/circulation/pull/506
both conflict with the work on this branch and are causing the tests to fail. 

Some documentation on testing this would have been really helpful. 

Here is what I did to try running the tests against this branch:
```
docker run --name test-opensearch -d --rm -p 9006:9200 -e "discovery.type=single-node" -e "plugins.security.disabled=true" "opensearchproject/opensearch:1"
docker exec test-opensearch opensearch-plugin -s install analysis-icu
docker restart test-opensearch
export SIMPLIFIED_TEST_ELASTICSEARCH=http://localhost:9006
```

That gave me a local opensearch instance to be able to do some testing against.

**Comment by jonathangreen** [comment]
Backing out the merge of main, I was able to run the unit tests against this branch. I think based on needing to resolve merge conflicts, and my requests above to allow us to run this against both `elasticsearch` and `opensearch` I'm going to stop this review here for now until some more work is done on this branch.

**Comment by RishiDiwanTT** [comment]
> So it looks like even though main merged in cleanly, some of the work we've done in:
> https://github.com/ThePalaceProject/circulation/pull/453
> https://github.com/ThePalaceProject/circulation/pull/506
> both conflict with the work on this branch and are causing the tests to fail.

Fixed the issues, both ES 7 and OS 1.2 have this breaking change (one document type per index)
I walked back the merge and did a rebase instead

**Comment by RishiDiwanTT** [comment]
> A couple questions that come up as I'm going through this:
> 
> * Do we need to update any libraries to be compatible with Opensearch? I see we are still using the elasticsearch libraries
> * Why are we targeting Opensearch 1.2.0? It looks like the most recent Opensearch version is 1.3.5

1. For the current build of 1.2 we do not have to change over any libraries, opensearch 2+ will need it's own libraries I believe
2. At the time of the work (26th July) AWS only had 1.2 available, as of [28th July](https://aws.amazon.com/about-aws/whats-new/2022/07/amazon-opensearch-service-supports-version-1-3/) it has 1.3 available, we can do another pass on this to upgrade it

**Comment by RishiDiwanTT** [comment]
> * Update this PR so the choice of running ES or OS is a configuration setting. You can either configure your site to work with Elasticsearch or Opensearch.

Are we talking about an upgraded ES 7.x or keeping it at the current ES 6.8.6?

**Comment by jonathangreen** [comment]
> Are we talking about an upgraded ES 7.x or keeping it at the current ES 6.8.6?

Keeping ES 6.8.6. That way we can merge this and keep sites running on ES 6.8.6, tackling the the upgrade to Opensearch slowly across all of our deployments.

**Comment by jonathangreen** [comment]
@RishiDiwanTT I tried to do some testing with this by upgrading my development environment to this code. My environment already had ES 6.8.6 running. Immediatly when I start the app I get this error: 

```
Traceback (most recent call last):
  File "/Users/jgreen/.pyenv/versions/3.8.1/envs/circulation-3/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/jgreen/.pyenv/versions/3.8.1/envs/circulation-3/lib/python3.8/site-packages/flask/app.py", line 1945, in full_dispatch_request
    self.try_trigger_before_first_request_functions()
  File "/Users/jgreen/.pyenv/versions/3.8.1/envs/circulation-3/lib/python3.8/site-packages/flask/app.py", line 1993, in try_trigger_before_first_request_functions
    func()
  File "/Users/jgreen/projects/libsimple/palace/circulation/api/routes.py", line 31, in initialize_circulation_manager
    app.manager = CirculationManager(app._db)
  File "/Users/jgreen/projects/libsimple/palace/circulation/api/controller.py", line 154, in __init__
    self.load_settings()
  File "/Users/jgreen/projects/libsimple/palace/circulation/core/util/log.py", line 24, in wrapper
    value = fn(*args, **kwargs)
  File "/Users/jgreen/projects/libsimple/palace/circulation/api/controller.py", line 244, in load_settings
    self.setup_external_search()
  File "/Users/jgreen/projects/libsimple/palace/circulation/api/controller.py", line 357, in setup_external_search
    self._external_search = self.setup_search()
  File "/Users/jgreen/projects/libsimple/palace/circulation/api/controller.py", line 415, in setup_search
    search = ExternalSearchIndex(self._db)
  File "/Users/jgreen/projects/libsimple/palace/circulation/core/external_search.py", line 288, in __init__
    self.set_works_index_and_alias(_db)
  File "/Users/jgreen/projects/libsimple/palace/circulation/core/external_search.py", line 320, in set_works_index_and_alias
    self._update_index_mapping()
  File "/Users/jgreen/projects/libsimple/palace/circulation/core/external_search.py", line 337, in _update_index_mapping
    current_mapping: Dict = _properties(
  File "/Users/jgreen/projects/libsimple/palace/circulation/core/external_search.py", line 334, in _properties
    mapping = mapping[self.work_document_type]
KeyError: 'works-type'
```

**Comment by jonathangreen** [comment]
I also gave this a try with an OS configuration instead, and I am getting a similar `KeyError` when trying to access the CM in that case. 😟

**Comment by jonathangreen** [comment]
Looking over this code, I think that the choice between ES and OS should be two different external integrations. I know I mentioned a configuration setting before, I should have been more precise in my language. I think different external integrations makes sense, since there is no migration path between the two. You would have to delete one external integration and add another one for OS.

**Comment by RishiDiwanTT** [comment]
> Looking over this code, I think that the choice between ES and OS should be two different external integrations. I know I mentioned a configuration setting before, I should have been more precise in my language. I think different external integrations makes sense, since there is no migration path between the two. You would have to delete one external integration and add another one for OS.

This architecture was my initial attempt at the implementation, however as I got into the actual implementation a quick search told me there were > 150 places the `ExternalSearchIndex` class was being imported and used as-is. Changing this to a dynamically changing class or class factory would have broken functionality in several unrelated workflows which seemed like a bad idea.

**Comment by RishiDiwanTT** [comment]
> @RishiDiwanTT I tried to do some testing with this by upgrading my development environment to this code. My environment already had ES 6.8.6 running. Immediatly when I start the app I get this error:
> 
> ```
> Traceback (most recent call last):
>   File "/Users/jgreen/.pyenv/versions/3.8.1/envs/circulation-3/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
>     response = self.full_dispatch_request()
>   File "/Users/jgreen/.pyenv/versions/3.8.1/envs/circulation-3/lib/python3.8/site-packages/flask/app.py", line 1945, in full_dispatch_request
>     self.try_trigger_before_first_request_functions()
>   File "/Users/jgreen/.pyenv/versions/3.8.1/envs/circulation-3/lib/python3.8/site-packages/flask/app.py", line 1993, in try_trigger_before_first_request_functions
>     func()
>   File "/Users/jgreen/projects/libsimple/palace/circulation/api/routes.py", line 31, in initialize_circulation_manager
>     app.manager = CirculationManager(app._db)
>   File "/Users/jgreen/projects/libsimple/palace/circulation/api/controller.py", line 154, in __init__
>     self.load_settings()
>   File "/Users/jgreen/projects/libsimple/palace/circulation/core/util/log.py", line 24, in wrapper
>     value = fn(*args, **kwargs)
>   File "/Users/jgreen/projects/libsimple/palace/circulation/api/controller.py", line 244, in load_settings
>     self.setup_external_search()
>   File "/Users/jgreen/projects/libsimple/palace/circulation/api/controller.py", line 357, in setup_external_search
>     self._external_search = self.setup_search()
>   File "/Users/jgreen/projects/libsimple/palace/circulation/api/controller.py", line 415, in setup_search
>     search = ExternalSearchIndex(self._db)
>   File "/Users/jgreen/projects/libsimple/palace/circulation/core/external_search.py", line 288, in __init__
>     self.set_works_index_and_alias(_db)
>   File "/Users/jgreen/projects/libsimple/palace/circulation/core/external_search.py", line 320, in set_works_index_and_alias
>     self._update_index_mapping()
>   File "/Users/jgreen/projects/libsimple/palace/circulation/core/external_search.py", line 337, in _update_index_mapping
>     current_mapping: Dict = _properties(
>   File "/Users/jgreen/projects/libsimple/palace/circulation/core/external_search.py", line 334, in _properties
>     mapping = mapping[self.work_document_type]
> KeyError: 'works-type'
> ```

This.. was a typo in the work_document_type that got introduced after ES was deleted and re-added.
It wasn't caught during my local testing because I blew away and re-indexed the search indexes several times.

**Comment by RishiDiwanTT** [comment]
> I also gave this a try with an OS configuration instead, and I am getting a similar `KeyError` when trying to access the CM in that case. worried

Could you try with both again. Changing the search version means the CM needs a restart.

File: core/testing.py
Comment (jonathangreen) [reply core/testing.py:1204]: We could maybe handle this on the tox side, and run the tests with an `es` and an `os` factor and select the tests to run by pytest mark. Might be a bit easier to understand whats going on.

**Review by jonathangreen — COMMENTED** [review]
This is looking good. Just a couple more minor comments there. And some thoughts about how we are running tests. I'd like a bit of a cleaner way to run the tests, but I don't have a lot of ideas there.

I setup a test environment with a clone of one of our production environments, and I was able to run it with no problem under ES 6.8, and then I setup a container with OS 1.3 and I was able to reindex everything, and then do searches against the OS index. I compared the results I got on OS and ES, and they were the same.

File: core/testing.py
Comment (RishiDiwanTT) [reply core/testing.py]: Updated. This is much cleaner, and is probably easier to understand too.

**Comment by RishiDiwanTT** [comment]
@jonathangreen bump! This has been rebased with the fixture changes

**Comment by RishiDiwanTT** [comment]
@jonathangreen @tdilauro 
Rebased again
https://media.giphy.com/media/59fYAHs0Y4llMQ7Lnh/giphy.gif

**Comment by jonathangreen** [comment]
@RishiDiwanTT I pushed a commit here, to use parameterized fixtures for these tests since @io7m fixture PR has landed. I think that is an easier to understand solution then what we were talking about before.

File: tests/core/test_external_search.py
Comment (jonathangreen) [inline tests/core/test_external_search.py:1311]: Did you mean to add this test? Or did it just come in by accident when rebasing these changes?

**Review by jonathangreen — COMMENTED** [review]
@RishiDiwanTT I pushed a commit here, to use parameterized fixtures for these tests since @io7m fixture PR has landed. I think that is an easier to understand solution then what we were talking about before. Can you take a look and see if my changes look okay to you? 

After all the rebasing I need to try running this again, to make sure things are still working correctly before I approve it. I think it is good to go though. Will approve once I test.

**Review by jonathangreen — APPROVED** [review]
Tested with ES and OS locally, and it all seemed good.

File: tests/core/test_external_search.py
Comment (RishiDiwanTT) [reply tests/core/test_external_search.py:1311]: This is added by me in September, this test was added because I had to make changes to the `doc_type` for `ExternalSearch.remove_work`

**Pull request merged**
